### ISSUE TITLE
False positive in generic constraint, seems to be due to unresolved forward reference

### ISSUE THREAD

**Issue by alwaysmpe** [microsoft/pyright#9641]
**Describe the bug**
I'm using a `TypeAliasType` to define the type of values in a tuple. The alias requires at least 2 values and can be multiple possible value types (below I've simplified to just one). I think this type forward reference isn't being resolved when the tuple alias is used. However, if an unused forward reference to the type value is added before the tuple alias the false positive goes away.

**Code or Screenshots**
```python
# type _OrFR = OrSeq # Uncommenting this line removes the error but this alias is unused.
type _OrVT =  OrSeq
type _OrT = tuple[_OrVT, _OrVT, *tuple[_OrVT, ...]]

class Base:
    pass

class Sequence[SequenceT: tuple[Base | str, ...]](Base):
    pass

class OrSeq(Sequence[_OrT]):
    # Type "_OrT" cannot be assigned to type variable "SequenceT@Sequence"
    #   Type "_OrT" is not assignable to upper bound "tuple[Base | str, ...]" for type variable "SequenceT@Sequence"
    #     "tuple[_OrVT, _OrVT, *tuple[_OrVT, ...]]" is not assignable to "tuple[Base | str, ...]"
    #       Tuple entry 1 is incorrect type
    #         Type "_OrVT" is not assignable to type "Base | str"
    #           Type "_OrVT" is not assignable to type "Base"
    #           Type "_OrVT" is not assignable to type "str"  (reportInvalidTypeArguments)
    pass
```
The error says that `_OrVT` is not assignable to type `Base | str`, but `_OrVT` is a forward reference to a subclass of `Base`.

**VS Code extension or command-line**
[pyright playground latest](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=MQAgLgngDgpiD6B5ATgMQEogLwhQZRgEcRQBVAOwGMB7AW1pnLAEtyBzcAC2YGcQAbVnGQxa1AG4w%2BYTnBjJk1ZCABGAVzBdeIAIaCdfbWvJqeMACYA6AFCRYCFADUAKthC5kBQrehwkyVxwwNSh%2BGABtfxcAGgdkGJAAKmDQiKjnWMssgF1s62tKfgM%2BACEDGAAua3d3KGL8wuKQLzVGSgiWtphnCvAQsPCysxAAHxAeMGRMnOyACiGYAEoqmpA6nh4Goo2PL1nOqjSUZ2zl6prQZ18QACJ-ZxuQSh1ycmpNFThi5jZyC3BqOBruIdMhmDoVGFbgd2s4AAIwmA3c7uUDuK72O7HR7aN6ab6-CFQsCAkKwZQqajGcy3FIDBajcaTaaWbKPABmSiB9hBYKJcBuiPhiORqzRNRudKO8QycQSyX60oSWVZbJAuPeug2P3I-IBtMVg3KjImUxAKrZKJIqxqzkVIEYkwgIAAjOrDFQlCJKJo7DAreKbSAMQL0ji%2BHitTwdXqSdyBQyxqbRRcgzaQ7cw%2B6QJGCbrIXA437bgsU6i06sM1iZeGc5q87HAcWbsn3LMRFAlGAAJLkEGCcwhgCCyDYagYTB4iyt602QA)

**Comment by erictraut** [comment]
You have created a circular reference here. The class `OrSeq` derives from `Sequence[_OrT]`, but `_OrT` is defined in terms of `OrSeq`. 

Pyright will attempt to resolve such a circularity, but the resulting behavior (e.g. which errors are reported) will depend where in the cycle it starts to evaluate types. 

You'll need to eliminate this cycle in your code if you want this to type check.

Closing because I don't consider this a bug in pyright.

**Comment by alwaysmpe** [comment]
Fair enough, thanks for your explanation. I think the best work around I've found is to use the type alias in the constraint and that triggers evaluation at the right point in the cycle (or alternatively the alias doesn't need to be evaluated to verify that the constraint is met).
